drivers/flash/soc_flash_b91.c

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2021 Telink Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT telink_b91_flash_controller
+#define FLASH_SIZE DT_REG_SIZE(DT_INST(0, soc_nv_flash))
+
+#include "flash.h"
+#include <device.h>
+#include <drivers/flash.h>
+
+
+/* driver data structure */
+struct flash_b91_data {
+	struct k_sem write_lock;
+};
+
+/* driver parameters structure */
+static const struct flash_parameters flash_b91_parameters = {
+	.write_block_size = DT_PROP(DT_INST(0, soc_nv_flash), write_block_size),
+	.erase_value = 0xff,
+};
+
+
+/* Check for correct offset and length */
+static bool flash_b91_is_range_valid(off_t offset, size_t len)
+{
+	/* check for min value */
+	if ((offset < 0) || (len < 1)) {
+		return false;
+	}
+
+	/* check for max value */
+	if ((offset + len) > FLASH_SIZE) {
+		return false;
+	}
+
+	return true;
+}
+
+/* API implementation: driver initialization */
+static int flash_b91_init(const struct device *dev)
+{
+	struct flash_b91_data *dev_data = dev->data;
+
+	k_sem_init(&dev_data->write_lock, 1, 1);
+
+	return 0;
+}
+
+/* API implementation: erase */
+static int flash_b91_erase(const struct device *dev, off_t offset, size_t len)
+{
+	int page_nums = len / PAGE_SIZE;
+	struct flash_b91_data *dev_data = dev->data;
+
+	/* check for valid length */
+	if (!len) {
+		return 0;
+	}
+
+	/* check for valid range */
+	if (!flash_b91_is_range_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	/* Erase can be done only by pages */
+	if (((offset % PAGE_SIZE) != 0) || ((len % PAGE_SIZE) != 0)) {
+		return -EINVAL;
+	}
+
+	/* take semaphore */
+	if (k_sem_take(&dev_data->write_lock, K_NO_WAIT)) {
+		return -EACCES;
+	}
```

Comment:
I am not sure whether this is needed at the level of the driver.
The flash_erase is blocking call, which means that calling erase in parallel, or write, will not be possible till current operation completes.
I think that the upper level, file system, or app, that performs access to the flash should make sure to not try to do two operations at the same offset;  fs or app will have to take care of avoiding parallel operations on the same offset, to keep structures it writes valid, which means that it may do some access locks anyway, this makes these locks redundant.

Reply:
It makes sense.
However, I prefer to live this verification because of safety reasons since it is present in all other vendor's flash drivers.

Reply:
OK.
Someday we should probably make it Kconfig option to enable this checks, as in most cases this is just a waste of a call.